### ISSUE TITLE
Semiautomatic determination of room ids

### ISSUE THREAD

**Issue by gothma** [nicoh88/homebridge-xiaomi-roborock-vacuum#149]
Hi,

first thank you for this plugin, really nice that I can now voice control my vacuum.

I have a roborock S5 and stumpled upon a way to retrieve the room id without trying out. Maybe it‘s helpful and become a new feature. I can also start working on a pull request, if you are interested.

Basically I set up a timer in the app for room cleaning, but the manually select all the rooms. The I call get_timer via miio and in the newly created timer the segment ids are included. The order of the segment ids is the order in which they were selected in the app. So now I immediately know all the segment ids and even which id is which without trying out.

I thought the autoroom function could be extended, so that if no rooms could be retrieved but a certain timer is set, the segments from this timer are auto generated. Let me know what you think, I could get working on it.

I can’t speak for other roborock versions, but I can imagine this could also work for others which have unnameable rooms.

**Comment by afharo** [comment]
Wow! That looks like it could work!! I'd love to see that in a PR if you are willing to do that! 🙏 

Please, remember to add a bit of documentation about this feature in the README so people can find out :)

**Comment by gothma** [comment]
Yeah I’m happy to do it. I’ve created a draft PR. Retrieving the roomIds from the robot’s timer seems to work fine, yet the services are not created. Can someone have a look?

The logs show that createRoom is called in the same way as when manually defining the rooms. Yet the services do not show up in HomeKit or the Homebridge UI. What I noticed, is that for the manual way the rooms are created in initialiseServices, before the robot is contacted. This is noticable from the undefined in the log message during createRoom. For autoroom they are created after the robot is connected, as they need to communicate with the robot.

So my guess is that the rooms are created too late, after some important initialization step. This is is weird because then the preexisting autoroom function should’ve also failed in creating the rooms, as it was called aswell after the initialization of the services.

Log with config `rooms: [...], autoroom: false`
```
[4/16/2020, 5:34:59 PM] [Staubsauger] Initializing XiaomiRoborockVacuum accessory...
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Wohnzimmer (18)
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Esszimmer (20)
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Küche (19)
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Flur (17)
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Bad (16)
[4/16/2020, 5:34:59 PM] [Staubsauger] DEB getDevice | Discovering vacuum cleaner
[4/16/2020, 5:34:59 PM] [Staubsauger] DEB getServices | undefined
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | Connected to: 192.168.3.164
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | Model: roborock.vacuum.s5
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | State: charging
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | FanSpeed: 101
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | BatteryLevel: 100
[4/16/2020, 5:35:00 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:35:00 PM] [Staubsauger] INF getSerialNumber | roborock.vacuum.s5 | Serial Number is 
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | Serialnumber: 
[4/16/2020, 5:35:00 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:35:00 PM] [Staubsauger] INF getFirmware | roborock.vacuum.s5 | Firmwareversion is 3.5.7_002008
[4/16/2020, 5:35:00 PM] [Staubsauger] STA getDevice | Firmwareversion: 3.5.7_002008
[4/16/2020, 5:35:00 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:35:00 PM] [Staubsauger] DEB getState | roborock.vacuum.s5 | State {"batteryLevel":100,"charging":true,"cleaning":false,"fanSpeed":101}
[4/16/2020, 5:35:00 PM] [Staubsauger] MON changedCleaning | roborock.vacuum.s5 | CleaningState is now false
[4/16/2020, 5:35:00 PM] [Staubsauger] INF changedCleaning | roborock.vacuum.s5 | Cleaning is OFF.
[4/16/2020, 5:35:00 PM] [Staubsauger] MON changedCharging | roborock.vacuum.s5 | ChargingState is now true
[4/16/2020, 5:35:00 PM] [Staubsauger] INF changedCharging | roborock.vacuum.s5 | Charging is active
[4/16/2020, 5:35:00 PM] [Staubsauger] MON changedSpeed | roborock.vacuum.s5 | FanSpeed is now 101%
[4/16/2020, 5:35:00 PM] [Staubsauger] INF changedSpeed | roborock.vacuum.s5 | Speed was changed to 101% (Quiet), for HomeKit 38%
[4/16/2020, 5:35:00 PM] [Staubsauger] DEB changedBattery | roborock.vacuum.s5 | BatteryLevel 100%
```

Log with `rooms: false, autoroom: [...]`
```
[4/16/2020, 5:34:59 PM] [Staubsauger] Initializing XiaomiRoborockVacuum accessory...                                                        
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Wohnzimmer (18)                                                     
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Esszimmer (20)                                                      
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Küche (19)                                                          
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Flur (17)                                                           
[4/16/2020, 5:34:59 PM] [Staubsauger] INF createRoom | undefined | Room Bad (16)                                                            
[4/16/2020, 5:34:59 PM] [Staubsauger] DEB getDevice | Discovering vacuum cleaner                                                            
[4/16/2020, 5:34:59 PM] [Staubsauger] DEB getServices | undefined
[4/16/2020, 5:43:18 PM] [Staubsauger] DEB getServices | undefined
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | Connected to: 192.168.3.164             
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | Model: roborock.vacuum.s5               
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | State: charging
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | FanSpeed: 101
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | BatteryLevel: 100                       
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:43:19 PM] [Staubsauger] INF getRoomList | roborock.vacuum.s5 | Room IDs are 18,20,17,19,16
[4/16/2020, 5:43:19 PM] [Staubsauger] INF createRoom | roborock.vacuum.s5 | Room Wohnzimmer (18)
[4/16/2020, 5:43:19 PM] [Staubsauger] INF createRoom | roborock.vacuum.s5 | Room Esszimmer (20)[4/16/2020, 5:43:19 PM] [Staubsauger] INF createRoom | roborock.vacuum.s5 | Room Flur (17)    
[4/16/2020, 5:43:19 PM] [Staubsauger] INF createRoom | roborock.vacuum.s5 | Room Küche (19)   
[4/16/2020, 5:43:19 PM] [Staubsauger] INF createRoom | roborock.vacuum.s5 | Room Bad (16)     
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:43:19 PM] [Staubsauger] INF getSerialNumber | roborock.vacuum.s5 | Serial Number
is
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | Serialnumber:
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:43:19 PM] [Staubsauger] INF getFirmware | roborock.vacuum.s5 | Firmwareversion is 3.5.7_002008
[4/16/2020, 5:43:19 PM] [Staubsauger] STA getDevice | Firmwareversion: 3.5.7_002008           
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB ensureDevice | roborock.vacuum.s5 | The socket is still on. Reusing it.
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB getState | roborock.vacuum.s5 | State {"batteryLevel":100,"charging":true,"cleaning":false,"fanSpeed":101}                                         
[4/16/2020, 5:43:19 PM] [Staubsauger] MON changedCleaning | roborock.vacuum.s5 | CleaningState is now false
[4/16/2020, 5:43:19 PM] [Staubsauger] INF changedCleaning | roborock.vacuum.s5 | Cleaning is OFF.
[4/16/2020, 5:43:19 PM] [Staubsauger] MON changedCharging | roborock.vacuum.s5 | ChargingState
is now true
[4/16/2020, 5:43:19 PM] [Staubsauger] INF changedCharging | roborock.vacuum.s5 | Charging is active
[4/16/2020, 5:43:19 PM] [Staubsauger] MON changedSpeed | roborock.vacuum.s5 | FanSpeed is now 101%
[4/16/2020, 5:43:19 PM] [Staubsauger] INF changedSpeed | roborock.vacuum.s5 | Speed was changed to 101% (Quiet), for HomeKit 38%
[4/16/2020, 5:43:19 PM] [Staubsauger] DEB changedBattery | roborock.vacuum.s5 | BatteryLevel 100%
```

**Comment by afharo** [comment]
> So my guess is that the rooms are created too late, after some important initialization step.

You are totally in the right direction! Homebridge does not allow registering services after the initialization. But I believe @c-o-m-m-a-n-d-e-r found [a hack to hold the registration for a bit](https://github.com/nicoh88/homebridge-xiaomi-roborock-vacuum/blob/master/index.js#L853-L854). I don't know if this could help here.

There is another way in homebridge to register services dynamically, but we'd need to rework this plugin to convert it into that. Maybe something to do in the future :)

**Comment by gothma** [comment]
Okay I think I begin to understand the problem. Thanks for the delay tip. Unfortunately it also delays getDevice, so the rooms are still created after getServices is called. So that did not work for me.

So I'll definetly change the debug output so that at least one can copy the mapping from the log to the config.

One botch idea that I had: I can create named rooms with incrementing ids in the constructor and lookup the actual roomId in the callback. What do you think, would that be ok?

**Comment by afharo** [comment]
Yeah! That seems like a valid approach to me:
We can register the services (since we no the room names beforehand from the config) and assign the roomId to the services once we get them.
👌

**Comment by gothma** [comment]
Alright I‘ve updated the PR: A json object is now printed to the log which can be directly copied to the config. Also I‘ve changed the room creation so that I can change ids afterwards. Feels hackish to me, let me know what you think. But with this room cleaning now works with my Roborock S5.

**Comment by gothma** [comment]
Okay, did the last changes and it still works. I’ve tested both the autoroom function and the manually declared rooms with my S5.

**Comment by afharo** [comment]
Published as `0.7.6`. Thank you kindly!
